test/langtools/tools/javac/launcher/ModuleSourceLauncherTests.java

```diff
@@ -220,4 +220,104 @@ public static void main(String... args) {
                 () -> assertTrue(err.isEmpty())
         );
     }
+
+    @Test
+    void testServiceLoading(@TempDir Path base) throws Exception {
+        var packageFolder = Files.createDirectories(base.resolve("p"));
+        var mainFile = Files.writeString(packageFolder.resolve("Main.java"),
+                """
+                package p;
+
+                import java.util.ServiceLoader;
+                import java.util.spi.ToolProvider;
+
+                class Main {
+                    public static void main(String... args) throws Exception {
```

Comment:
Stylistically, it can be neater to abstract the use of `System.out` to avoid repeated use of the field.

Reply:
The intention here was to use a few idiomatic lines known from many simple Java programs. "Copy and paste" turned into a larger pattern then expected: will use a shorter form in the future. Either using a field or `IO.println()` in combination with an `import static java.io.IO.println;`